presto-orc/src/main/java/com/facebook/presto/orc/block/MapBlockReader.java

```diff
@@ -81,21 +83,17 @@ public boolean readNextValueInto(BlockBuilder builder, boolean skipNull)
         }
 
         long length = lengthStream.next();
-        BlockBuilder currentBuilder = VARBINARY.createBlockBuilder(new BlockBuilderStatus(), Ints.checkedCast(length));
+        BlockBuilder currentBuilder = new InterleavedBlockBuilder(typeParameters, new BlockBuilderStatus(), Ints.checkedCast(length));
         for (int i = 0; i < length; i++) {
-            if (keyReader.readNextValueInto(currentBuilder, true)) {
-                valueReader.readNextValueInto(currentBuilder, false);
+            if (keyReader.readNextValueInto(keyType, currentBuilder, true)) {
+                valueReader.readNextValueInto(valueType, currentBuilder, false);
             }
             else {
                 valueReader.skip(1);
             }
         }
+        type.writeObject(builder, currentBuilder.build());
```

Comment:
also if you use an entry writer here and in the list decoder, then you can remove `Type` from `readNextValueInto`